File: AuctionMarket/src/main/java/com/portfolio/auctionmarket/auth/service/AuthService.java
package com.portfolio.auctionmarket.auth.service;

import com.portfolio.auctionmarket.auth.dto.LoginRequest;
import com.portfolio.auctionmarket.auth.dto.SecurityUser;
import com.portfolio.auctionmarket.auth.dto.TokenResponse;
import com.portfolio.auctionmarket.domain.user.entity.User;
import com.portfolio.auctionmarket.domain.user.entity.UserStatus;
import com.portfolio.auctionmarket.domain.user.repository.UserRepository;
import com.portfolio.auctionmarket.global.config.JwtProperties;
import com.portfolio.auctionmarket.global.error.CustomException;
import com.portfolio.auctionmarket.global.error.ErrorCode;
import lombok.RequiredArgsConstructor;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.LockedException;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final JwtService jwtService;
    private final UserRepository userRepository;
    private final JwtProperties jwtProperties;
    private final RefreshTokenService refreshTokenService;
    private final AuthenticationManager authenticationManager;
    private final RedisTemplate<String, String> redisTemplate;

    @Transactional
    public TokenResponse login(LoginRequest request) {

        try {
            Authentication authentication = authenticationManager.authenticate(new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword()));

            SecurityUser securityUser = (SecurityUser) authentication.getPrincipal();

            String accessToken = jwtService.generateAccessToken(
                    securityUser.getUserId(),
                    securityUser.getUsername(),
                    securityUser.getNickname(),
                    securityUser.getUser().getRole().name()
            );
            String refreshToken = jwtService.generateRefreshToken(securityUser.getUserId());

            refreshTokenService.saveRefreshTokenBidirectional(securityUser.getUserId(), refreshToken);

            return TokenResponse.builder()
                    .accessToken(accessToken)
                    .refreshToken(refreshToken)
                    .expiresIn(jwtProperties.getAccessTokenExpiration() / 1000)
                    .tokenType("Bearer")
                    .build();
        } catch (LockedException e){
            throw new CustomException(ErrorCode.SUSPENDED_USER, "정지된 사용자 입니다.");
        }
    }

    public TokenResponse refreshAccessToken(String refreshToken) {
        jwtService.validateToken(refreshToken);

        Long userId = refreshTokenService.getUserIdByToken(refreshToken);

        if (userId == null) {
            throw new CustomException(ErrorCode.TOKEN_NOT_FOUND);
        }

        String redisRefreshToken = redisTemplate.opsForValue().get("refresh:user:" + userId);

        if (redisRefreshToken == null || !redisRefreshToken.equals(refreshToken)) {
            throw new CustomException(ErrorCode.INVALID_TOKEN);
        }

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new CustomException(ErrorCode.USER_NOT_FOUND));

        String newAccessToken = jwtService.generateAccessToken(user.getUserId(), user.getEmail(), user.getNickname(), user.getRole().name());
        String newRefreshToken = jwtService.generateRefreshToken(user.getUserId());

        refreshTokenService.saveRefreshTokenBidirectional(user.getUserId(), newRefreshToken);

        return TokenResponse.builder()
                .accessToken(newAccessToken)
                .refreshToken(newRefreshToken)
                .expiresIn(jwtProperties.getAccessTokenExpiration() / 1000)
                .build();
    }

    public void logout(Long userId) {
        refreshTokenService.deleteRefreshToken(userId);
    }
}

File: AuctionMarket/auction-app-user/src/main/kotlin/org/example/auctioncommon/auth/service/AuthService.kt
package org.example.auctioncommon.auth.service

import org.example.auctioncommon.auth.dto.LoginRequest
import org.example.auctioncommon.auth.dto.SecurityUser
import org.example.auctioncommon.auth.dto.TokenResponse
import org.example.auctioncommon.domain.user.entity.User
import org.example.auctioncommon.domain.user.error.UserErrorCode
import org.example.auctioncommon.domain.user.repository.UserRepository
import org.example.auctioncommon.global.config.JwtProperties
import org.example.auctioncommon.global.error.CustomException
import org.example.auctioncommon.global.error.GlobalErrorCode
import org.springframework.data.repository.findByIdOrNull
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.LockedException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional


@Service
class AuthService(
    private val jwtService: JwtService,
    private val jwtProperties: JwtProperties,
    private val authProcessor: AuthProcessor,
    private val userRepository: UserRepository,
    private val refreshTokenService: RefreshTokenService,
    private val authenticationManager: AuthenticationManager,
) {


    @Transactional
    fun login(request: LoginRequest): TokenResponse {
        try {
            val authentication = authenticationManager.authenticate(
                UsernamePasswordAuthenticationToken(
                    request.email,
                    request.password
                )
            )

            val securityUser: SecurityUser? = authentication.principal as SecurityUser?

            val accessToken: String = jwtService.generateAccessToken(
                userId = securityUser?.userId,
                email = securityUser?.username.toString(),
                nickname = securityUser?.nickname.toString(),
                role = securityUser?.user?.role?.name.toString()
            )
            val refreshToken: String = jwtService.generateRefreshToken(securityUser?.userId ?: 0L)

            refreshTokenService.saveRefreshTokenBidirectional(securityUser?.userId ?: 0L, refreshToken)

            return TokenResponse(
                accessToken = accessToken,
                refreshToken = refreshToken,
                expiresIn = jwtProperties.accessTokenExpiration?.div(1000L)
            )
        } catch (e: LockedException) {
            throw CustomException(UserErrorCode.SUSPENDED_USER, "정지된 사용자 입니다.")
        }
    }

    fun refreshAccessToken(refreshToken: String): TokenResponse {
        jwtService.validateToken(refreshToken)

        val userId: Long =
            refreshTokenService.getUserIdByToken(refreshToken) ?: throw CustomException(GlobalErrorCode.TOKEN_NOT_FOUND)

        val redisRefreshToken: String? = refreshTokenService.redisRefreshToken(userId)

        authProcessor.validateRedisRefreshToken(redisRefreshToken, refreshToken)

        if (redisRefreshToken == null || redisRefreshToken != refreshToken) {
            throw CustomException(GlobalErrorCode.INVALID_TOKEN)
        }

        val user: User = userRepository.findByIdOrNull(userId)
            ?: throw CustomException(UserErrorCode.USER_NOT_FOUND)

        val newAccessToken: String =
            jwtService.generateAccessToken(userId, user.email, user.nickname, user.role.name)
        val newRefreshToken: String = jwtService.generateRefreshToken(user.userId)

        refreshTokenService.saveRefreshTokenBidirectional(userId, newRefreshToken)

        return TokenResponse(
            accessToken = newAccessToken,
            refreshToken = newRefreshToken,
            expiresIn = jwtProperties.accessTokenExpiration?.div(1000L)
        )
    }

    fun logout(userId: Long) {
        refreshTokenService.deleteRefreshToken(userId)
    }
}
